token/program-2022/src/extension/reallocate.rs

```diff
@@ -83,8 +84,23 @@ pub fn process_reallocate(
     )?;
 
     // unpack to set account_type, if needed
-    let mut token_account = token_account_info.data.borrow_mut();
-    set_account_type::<Account>(&mut token_account)?;
+    let mut token_account_data = token_account_info.data.borrow_mut();
+    set_account_type::<Account>(&mut token_account_data)?;
+
+    // sync the rent exempt reserve for native accounts
+    let mut token_account = StateWithExtensionsMut::<Account>::unpack(&mut token_account_data)?;
+    if token_account.base.is_native.is_some() {
+        let new_amount = token_account_info
+            .lamports()
+            .checked_sub(new_minimum_balance)
+            .ok_or(TokenError::Overflow)?;
+        if new_amount < token_account.base.amount {
```

Comment:
Currently, we only progress past this check if the native account is "empty", ie. only holds rent-exempt reserve and base amount = 0. I assume that is not your intent, is it?

To fix, I think we need to primarily edit L76, but I can't suggest there. Something like:
```
    let lamports_to_be_subtracted = if let Some(amount) = native_amount {
        token_account_info.lamports().saturating_sub(amount)
    } else {
        token_account_info.lamports()
    };
    let lamports_diff = new_minimum_balance.saturating_sub(lamports_to_be_subtracted);
```
where `native_amount` is returned in the block on L37 so we don't have to unpack the account again.
Something like:
```
    let (mut current_extension_types, native_amount) = {
        let token_account = token_account_info.data.borrow();
        let account = StateWithExtensions::<Account>::unpack(&token_account)?;
        Processor::validate_owner(
            program_id,
            &account.base.owner,
            authority_info,
            authority_info_data_len,
            account_info_iter.as_slice(),
        )?;
        let native_amount = if account.base.is_native() {
            Some(account.base.amount)
        } else {
            None
        };
        (account.get_extension_types()?, native_amount)
    };
```

Reply:
Ah right of course! Thanks for noticing this issue. In my head I thought it would just transfer the right amount of lamports in, which is totally wrong.

This seems like the most secure approach. We could also use the value stored in `is_native` when calculating the amount of lamports to transfer, but I think we'll be safer by not trusting that always.